CMake build system fixes

delcypher opened the pull request:
This PR contains several fixes to CMake build system.  Note build of apps is disabled because building them is still broken.

The main highlights are:

* A more sensible way of detecting and using LLVM (using its CMake Config file)
* Support for building Doxygen.
* Tutorials now build.

There are still lots of things that need to be done to bring the CMake build system to parity with the Makefile build system but hopefully we'll be able to remove it eventually because I don't think maintaining two build systems is a good idea.

abestephensg commented:
When I launch ccmake like this:
$ /Applications/CMake.app/Contents/bin/ccmake -GXcode ..

I get the following error:
CMake Error at CMakeLists.txt:14 (find_package):
   By not providing "FindLLVM.cmake" in CMAKE_MODULE_PATH this project has
   asked CMake to find a package configuration file provided by "LLVM", but
   CMake did not find one.

   Could not find a package configuration file provided by "LLVM" with any of
   the following names:

     LLVMConfig.cmake
     llvm-config.cmake

   Add the installation prefix of "LLVM" to CMAKE_PREFIX_PATH or set
   "LLVM_DIR" to a directory containing one of the above files.  If "LLVM"
   provides a separate development package or SDK, be sure it has been
   installed.

Can we check and catch this condition and print a less verbose and more helpful error message?  

In my case it was in <llvm install path>/share/llvm/cmake. Once I fixed that, the next error reported is:

CMake Error at CMakeLists.txt:91 (message):
   Could not find Doxygen.  Either install it or set WITH_DOCS to OFF

Could we set WITH_DOCS to OFF by default so that developers don't run into an error about an optional dependency?

I was able to successfully build all of the targets- and it's great to see the tutorials in the build as well!

abadams commented on README.md:
```diff
@@ -110,11 +110,14 @@ There is no 'make install' yet. If you want to make an install
     % cd build
     % LLVM_ROOT=/path/to/llvm3.5
-    % cmake -DLLVM_BIN=${LLVM_ROOT}/build/bin -DLLVM_INCLUDE="${LLVM_ROOT}/include;${LLVM_ROOT}/build/include" -DLLVM_LIB=${LLVM_ROOT}/build/lib -DLLVM_VERSION=35 ..
+    % cmake -DLLVM_DIR=${LLVM_ROOT}/build/share/cmake/
```
Doesn't this clobber the instructions for people who downloaded an llvm distribution?

delcypher replied:
@abadams What do you mean by clobber here?

If the user has a binary build they can also set ``LLVM_DIR`` to the directory containing ``LLVMConfig.cmake`` in the distribution they downloaded.

abadams replied:
I was concerned that these new instructions are not correct if you downloaded a binary distribution of llvm, however I see that the downloaded versions contain share/llvm/cmake/... too, so no worries.

abadams commented on Doxyfile.in:
```diff
@@ -0,0 +1,1552 @@
+# Doxyfile 1.6.2
```
Does this file exist in addition to Doxyfile? It seems to be a new copy of it.

delcypher replied:
It is a new copy :(

I really didn't want to do that but it is essential for out of source builds to work (if you take a look in ``Doxyfile.in`` you'll see there are a bunch of ``@`` substitutions in there).

I'd rather there wasn't two. I'd argue that eventually we'd want to get rid of the ``Makefile`` build system all together, when we can drop having two Doxyfiles.

zvookin replied:
"'I'd argue that eventually we'd want to get rid of the Makefile build
system all together"

That will not be happening unless we do something like Bazel support.

-Z-


On Thu, Jun 18, 2015 at 11:25 AM, Dan Liew <notifications@github.com> wrote:

> In Doxyfile.in
> <https://github.com/halide/Halide/pull/825#discussion_r32761840>:
>
> > @@ -0,0 +1,1552 @@
> > +# Doxyfile 1.6.2
>
> It is a new copy :(
>
> I really didn't want to do that but it is essential for out of source
> builds to work (if you take a look in Doxyfile.in you'll see there are a
> bunch of @ substitutions in there).
>
> I'd rather there wasn't two. I'd argue that eventually we'd want to get
> rid of the Makefile build system all together, when we can drop having
> two Doxyfiles.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825/files#r32761840>.
>

abadams replied:
<Charlton Heston cold-dead-hands meme.>

I find the Makefile many times easier to understand and maintain. CMake is
tolerated because it produces visual studio solutions, but we've lost many
man hours just trying to get it to do the same thing as something that was
easy in the Makefile.

From a Makefile's point of view your project is a DAG with some shell
scripts it needs to run to get from one node to another. If you don't rely
on default rules, it doesn't even know you're compiling something. Build
systems that attempt to understand your source code in some deeper way
usually shit all over themselves when you have generated source,
metaprogramming with multi-stage compilation, weird linkage options, etc.


On Thu, Jun 18, 2015 at 11:25 AM, Dan Liew <notifications@github.com> wrote:

> In Doxyfile.in
> <https://github.com/halide/Halide/pull/825#discussion_r32761840>:
>
> > @@ -0,0 +1,1552 @@
> > +# Doxyfile 1.6.2
>
> It is a new copy :(
>
> I really didn't want to do that but it is essential for out of source
> builds to work (if you take a look in Doxyfile.in you'll see there are a
> bunch of @ substitutions in there).
>
> I'd rather there wasn't two. I'd argue that eventually we'd want to get
> rid of the Makefile build system all together, when we can drop having
> two Doxyfiles.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825/files#r32761840>.
>

aam commented:
I'm running into problems running this on Windows:

```
C:\Users\me\Documents\halide\halide-del\build>"c:\Program Files (x86)\CMake\bin\cmake.exe" -DLLVM_DIR=c:\Users\me\Documents\halide\llvm-build\share\llvm\cmake -G "Visual
Studio 14 2015" ..
-- Looking for LLVM
CMake Warning (dev) at C:/Users/me/Documents/halide/llvm-build/share/llvm/cmake/LLVMExports.cmake:232 (set_target_properties):
  Syntax error in cmake code at

    C:/Users/me/Documents/halide/llvm-build/share/llvm/cmake/LLVMExports.cmake:233

  when parsing string

    LLVMObject;LLVMSupport;C:\Program Files (x86)\Microsoft Visual Studio 14.0\DIA SDK\lib\diaguids.lib

  Invalid escape sequence \P

  Policy CMP0010 is not set: Bad variable reference syntax is an error.  Run
  "cmake --help-policy CMP0010" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.
Call Stack (most recent call first):
  c:/Users/me/Documents/halide/llvm-build/share/llvm/cmake/LLVMConfig.cmake:174 (include)
  CMakeLists.txt:14 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found LLVM 3.7.0svn
-- LLVM_BIN:C:/Users/me/Documents/halide/llvm-build/$(Configuration)/bin
-- LLVM_INCLUDE:C:/Users/me/Documents/halide/llvm/include;C:/Users/me/Documents/halide/llvm-build/include
-- LLVM_VERSION:37
CMake Error at CMakeLists.txt:33 (if):
  Syntax error in cmake code at

    C:/Users/me/Documents/halide/halide-del/CMakeLists.txt:33

  when parsing string

    C:\Users\me\Documents\halide\llvm-build\$(Configuration)\bin\llvm-as.exe

  Invalid escape sequence \U
Call Stack (most recent call first):
  CMakeLists.txt:38 (check_tool_exists)


-- Configuring incomplete, errors occurred!
See also "C:/Users/me/Documents/halide/halide-del/build/CMakeFiles/CMakeOutput.log".
```

delcypher commented:
@aam I've just rebased and tried adding a fix for the second issue you saw

```
CMake Error at CMakeLists.txt:33 (if):
  Syntax error in cmake code at

    C:/Users/me/Documents/halide/halide-del/CMakeLists.txt:33
```

The first warning you see looks like a bug in CMake or LLVM. I need to see you LLVMExports.cmake file but I'm guessing that for some reason your build of LLVM links against ``C:\Program Files (x86)\Microsoft Visual Studio 14.0\DIA SDK\lib\diaguids.lib`` and the path is not written correctly, it needs to be a CMake style path.

Could you delete your copy of my branch and then pull it again and try building?

delcypher commented:
@zvookin @abadams 
https://github.com/delcypher/Halide/commit/6954e661b84e8394501647f3a40cafe7bf4efe9f 

teaches the Makefile build system to configure the ``Doxyfile.in`` file so now there is no need for two copies of the Doxyfile.

delcypher commented:
@aam I've patched one of the issues you saw that should properly fix one of the error messages you saw (check_tool_exists). I have [a patch for LLVM](http://lists.cs.uiuc.edu/pipermail/llvmdev/2015-June/086965.html) that might fix the issue with the broken ``LLVMExports.cmake`` created by LLVM but I can't test it as I don't have access to a Windows machine with Visual Studio. If you have the time feel free to test it.

delcypher commented:
@abestephensg 
> Can we check and catch this condition and print a less verbose and more helpful error message?

I've done this in https://github.com/delcypher/Halide/commit/9bca90619f165bf50ee28dbdd69df9a2b01406a5

> Could we set WITH_DOCS to OFF by default so that developers don't run into an error about an optional dependency?

This has been done in https://github.com/delcypher/Halide/commit/5447e08e5ae1de41532abc04c3aa30c34c8a9b15

aam commented:
@delcypher I gave llvm patch quick try and with it cmake complains that it can no longer find LLVM:

===
C:\Users\me\Documents\halide\halide-del\build>"c:\Program Files (x86)\CMake\bin\cmake.exe" -DLLVM_DIR=c:\Users\me\Documents\halide\llvm-build\share\llvm\cmake -G "Visual
Studio 14 2015" ..
-- The C compiler identification is MSVC 19.0.22816.0
-- The CXX compiler identification is MSVC 19.0.22816.0
-- Check for working C compiler using: Visual Studio 14 2015
-- Check for working C compiler using: Visual Studio 14 2015 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler using: Visual Studio 14 2015
-- Check for working CXX compiler using: Visual Studio 14 2015 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for LLVM
CMake Error at CMakeLists.txt:18 (message):
  Could not automatically find LLVM.  Try setting passing -DLLVM_DIR to CMake
  and setting it to the directory containing the LLVMConfig.cmakefile.  This
  file is available in the LLVM build tree under share/llvm/cmake/ if LLVM
  was built with CMake.  It is also available inLLVM's install tree


-- Configuring incomplete, errors occurred!
See also "C:/Users/me/Documents/halide/halide-del/build/CMakeFiles/CMakeOutput.log".

===

delcypher commented:
@aam Could you try reverting https://github.com/delcypher/Halide/commit/9bca90619f165bf50ee28dbdd69df9a2b01406a5

This was added add @abestephensg request but it looks like I broke something.

aam commented:
@delcypher I am not sure that is the problem because cmake started to worked fine when I reverted llvm patch.

delcypher commented:
@aam https://github.com/delcypher/Halide/commit/9bca90619f165bf50ee28dbdd69df9a2b01406a5 might be obscuring error messages that CMake would normally emit when building.

aam commented:
resetting to previous rev gives following error:
===
C:\Users\me\Documents\halide\halide-del\build>"c:\Program Files (x86)\CMake\bin\cmake.exe" -DLLVM_DIR=c:\Users\me\Documents\halide\llvm-build\share\llvm\cmake -G "Visual
Studio 14 2015" ..
-- Looking for LLVM
-- Found LLVM 3.7.0svn
-- LLVM_BIN:C:/Users/me/Documents/halide/llvm-build/Debug/bin
-- LLVM_INCLUDE:C:/Users/me/Documents/halide/llvm/include;C:/Users/me/Documents/halide/llvm-build/include
-- LLVM_VERSION:37
CMake Error at CMakeLists.txt:50 (message):
  Tool llvm-as not found at
  C:/Users/me/Documents/halide/llvm-build/Debug/bin/llvm-as.exe
Call Stack (most recent call first):
  CMakeLists.txt:54 (check_tool_exists)


-- Configuring incomplete, errors occurred!
See also "C:/Users/me/Documents/halide/halide-del/build/CMakeFiles/CMakeOutput.log".
===

I see Debug instead of Release in the path, which seems to be a problem.

delcypher commented:
Eurgh... well I guess that means my patch to LLVM is working (no errors doing ``find_package(LLVM)``). But my workaround to get the binary build directory in Halide isn't working

```
  # WORKAROUND: Read the location from an imported target to get LLVM's binary
  # directory. Using the ${LLVM_TOOLS_BINARY_DIR} variable is currently broken
  # when using Visual Studio (it contains the string $(Configuration) in the
  # path). We shouldn't read the location from the imported binary targets on
  # other platforms because if LLVM was built with the Autoconf/Makefile build
  # system they won't be available.
  get_target_property(LLVM_AS_PATH llvm-as LOCATION)
  get_filename_component(LLVM_BIN "${LLVM_AS_PATH}" DIRECTORY)
```

This code reads the ``LOCATION`` property from the ``llvm-as`` target. On my machine it is only set once

```
# Import target "llvm-as" for configuration "Release"
set_property(TARGET llvm-as APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(llvm-as PROPERTIES
  IMPORTED_LOCATION_RELEASE "/usr/local/google/home/liew/dev/halide/llvm/build/bin/llvm-as"
  )
```

but it seems for the visual studio build it is set multiple times (in ``LLVMExports.cmake``)

```
# Import target "llvm-as" for configuration "Debug"
set_property(TARGET llvm-as APPEND PROPERTY IMPORTED_CONFIGURATIONS DEBUG)
set_target_properties(llvm-as PROPERTIES
  IMPORTED_LOCATION_DEBUG "C:/Users/alexander.APRELEV/Documents/halide/llvm-build/Debug/bin/llvm-as.exe"
  )

...

# Import target "llvm-as" for configuration "Release"
set_property(TARGET llvm-as APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(llvm-as PROPERTIES
  IMPORTED_LOCATION_RELEASE "C:/Users/alexander.APRELEV/Documents/halide/llvm-build/Release/bin/llvm-as.exe"
  )

...

# Import target "llvm-as" for configuration "Release"
set_property(TARGET llvm-as APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(llvm-as PROPERTIES
  IMPORTED_LOCATION_RELEASE "C:/Users/alexander.APRELEV/Documents/halide/llvm-build/Release/bin/llvm-as.exe"
  )

...

# Import target "llvm-as" for configuration "RelWithDebInfo"
set_property(TARGET llvm-as APPEND PROPERTY IMPORTED_CONFIGURATIONS RELWITHDEBINFO)
set_target_properties(llvm-as PROPERTIES
  IMPORTED_LOCATION_RELWITHDEBINFO "C:/Users/alexander.APRELEV/Documents/halide/llvm-build/RelWithDebInfo/bin/llvm-as.exe"
  )

```

One workaround is to cycle through the different configurations and pick the first executable that exists. The other is to force Visual Studio users to tell us which configuration to use from the LLVM build tree (I prefer this). Thoughts?

delcypher commented:
Okay I think this is about ready. Now that I have access to Windows machine with Visual Studio fixing build issues has become a lot easier. The tutorials now build on Windows too :). Building the Apps is still disabled, that will be fixed at some point in the future.

Please note that I've rebased on master so if you had a copy of my branch you'll need to delete it and pull it again

@abadams Could you test again. I think most of it should be okay but  I'm under about https://github.com/delcypher/Halide/commit/3a86bcd9bfc692f91b7f63fee9e3a22e68b0ed09 This change seemed to be necessary when I rebased which I think it to do with recent changes to remove a dependency on the CUDA library.

@aam Feel free to take a look. I have the build working under Windows. You need to specify ``LLVM_BUILD_TYPE`` (usually ``DEBUG`` or ``RELEASE``) when building in addition to ``LLVM_DIR`` when building against LLVM built with Visual Studio.

I'd like get this merged soon I don't need to keep rebasing it.

abadams commented on tutorial/clock.h:
```diff
@@ -2,14 +2,16 @@
 #ifdef _WIN32
-extern "C" bool QueryPerformanceCounter(uint64_t *);
-extern "C" bool QueryPerformanceFrequency(uint64_t *);
+#include <Windows.h>
```
Why add the dependency on Windows.h here (given that you need to undef things).

delcypher replied:
Because the current version in master does not compile. I do not remember the exact error but the function prototypes defined originally did not match what was being used elsewhere and it caused a linking error.

abadams replied:
That's a good reason :)

abadams commented on Doxyfile.in:
```diff
@@ -631,7 +631,7 @@ EXCLUDE_SYMBOLS        =
 # the \include command).
 
-EXAMPLE_PATH           = tutorial
+EXAMPLE_PATH           = @CMAKE_SOURCE_DIR@
```
Shouldn't this include "tutorial"?

delcypher replied:
No because ``src/Mainpage.h`` uses ``example tutorial/lesson_01_basics.cpp``. Using ``tutorial`` caused lots of warnings to be emitted about this by Doxygen

abadams commented:
Does test/scripts/distrib_windows.ps1 need updating? That's how the nightlies are built on windows.

abadams commented:
Something seems messed up with the generator test dependencies. Everything works fine if I turn off test_generators. However, this also appears to be true in trunk, so it must be an existing error. Any ideas?

```
[ 89%] Generating acquire_release.a, acquire_release.h
cd /usr/local/google/home/abadams/projects/Halide-delcypher/build/test/scratch_acquire_release && /usr/local/google/home/abadams/projects/Halide-delcypher/build/bin/generator_acquire_release -g acquire_release -f acquire_release -o /usr/local/google/home/abadams/projects/Halide-delcypher/build/test/scratch_acquire_release target=host
cd /usr/local/google/home/abadams/projects/Halide-delcypher/build/test/scratch_acquire_release && /usr/bin/ar q acquire_release.a /usr/local/google/home/abadams/projects/Halide-delcypher/build/test/scratch_acquire_release/acquire_release.o
make[2]: *** [test/acquire_release.h] Error 1
make[2]: Leaving directory `/usr/local/google/home/abadams/projects/Halide-delcypher/build'
make[1]: *** [test/CMakeFiles/exec_generator_acquire_release_acquire_release.dir/all] Error 2
make[1]: Leaving directory `/usr/local/google/home/abadams/projects/Halide-delcypher/build'
make: *** [all] Error 2
```

delcypher commented:
@abadams Are you running on OSX? Because if so I'm having this issue too. I'm trying to track it down but the issue I'm seeing is that in the generated object file (``test/scratch_acquire_release/acquire_release.o``) the symbols are missing the ``_`` prefix (needed on OSX) resulting in linking errors.

abadams commented:
I'm running ubuntu.

On Tue, Jun 23, 2015 at 3:45 PM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams> Are you running on OSX? Because if
> so I'm having this issue too. I'm trying to track it down but the issue I'm
> seeing is that in the generated object file (
> test/scratch_acquire_release/acquire_release.o) the symbols are missing
> the _ prefix (needed on OSX) resulting in linking errors.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825#issuecomment-114665174>.
>

delcypher commented:
@abadams Update: I've built the released LLVM 3.6.1 from source and Halide builds without failing against that. So I guess it's a bug in upstream LLVM unrelated to this PR.

FYI I was using the following SVN revisions
```
LLVM: 240427
Clang: 240430
Compiler-rt: 240379
```

abadams commented:
It fails in the same way for me with llvm release 3.5, 3.6, and trunk. It
seems to think it's generating test/acquire_release.h, where what exists is
test/scratch_acquire_release/acquire_release.h

On Tue, Jun 23, 2015 at 3:52 PM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams> Update: I've built the released
> LLVM 3.6.1 from source and Halide builds without failing against that. So I
> guess it's a bug in upstream LLVM unrelated to this PR.
>
> FYI I was using the following SVN revisions
>
> LLVM: 240427
> Clang: 240430
> Compiler-rt: 240379
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825#issuecomment-114666299>.
>

delcypher commented:
Oh are you using ninja as your generator? Because I noticed that was broken. Using the makefile generator works right now.

delcypher commented:
@abadams https://github.com/delcypher/Halide/commit/aa2daeeaaf7ae9e2545caa6e86194a0be61390e9 modifies the Windows build script.

Good to go?

delcypher commented:
> Something seems messed up with the generator test dependencies. Everything works fine if I turn off test_generators. However, this also appears to be true in trunk, so it must be an existing error. Any ideas?

#832 addresses the issue

abadams commented:
Trying to test this on Windows, but LLVM COFF file generation seems borked right now (all the cross-compilation tests barfed overnight with segfaults inside llvm).

delcypher commented:
@abadams Presumably this an issue with LLVM trunk and not with my PR right?

abadams commented:
Correct. It's just making it hard to test the PR.

On Wed, Jun 24, 2015 at 4:29 PM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams> Presumably this an issue with LLVM
> trunk and not with my PR right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825#issuecomment-115045072>.
>

abadams commented:
For pnacl llvm, there needs to be a way to specify that the tools (clang, llvm-as) come from a different directory to the compiled llvm libraries. pnacl llvm's tools don't build on windows in msvc, but the libraries do, so I've scavenged the tools from elsewhere.

abadams commented:
e.g. if LLVM_BIN still works and overruled LLVM_DIR if present, that would fix the problem

delcypher commented:
@abadams. ``LLVM_BIN`` has been removed. However I could put it back. I could make it so that by default  it is set to the binaries of the LLVM build that CMake finds.

I'll take a look at this later today.

delcypher commented:
@abadams I've added support ``LLVM_BIN``

delcypher commented:
@abadams Just in case you were thinking of merging this in the next few hours please don't. I discovered a bug in LLVM last night that impacts this change. I'm working on a fix and will push it once it's done.

abadams commented:
Ok, I was currently testing it on Windows. Let me know when I should pull
and restart.

On Fri, Jun 26, 2015 at 9:15 AM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams> Just in case you were thinking of
> merging this in the next few hours please don't. I discovered a bug in LLVM
> last night that impacts this change. I'm working on a fix and will push it
> once it's done.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825#issuecomment-115744104>.
>

abadams commented:
Appears to work with that older known good llvm version. More to the point, pnacl llvm seems good now.

abadams commented:
btw, is the trunk llvm issue a segfault inside generating relocations for COFF files? That's what I'm getting on linux in the cross-compilation test with trunk llvm.

delcypher commented:
@abadams Okay. The fix (https://github.com/delcypher/Halide/commit/8df9e75e57e727f58e4b8ae44c0a82f9ebc1dbbf) is  in

delcypher commented:
@abadams No it's a different bug. It's a CMake issue. I'm [fixing it now](https://llvm.org/bugs/show_bug.cgi?id=23967)

By the way how do I actually run the tests when doing a CMake build? I don't see a nice target I can use to run them. I was thinking of maybe implementing this in a future PR.

abadams commented:
The powershell script just iterates over all the exes and runs them. Feel
free to submit a PR that adds a target. Or just use the Makefile...

On Fri, Jun 26, 2015 at 12:58 PM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams> No it's a different bug. It's a
> CMake issue. I'm fixing it now
> <https://llvm.org/bugs/show_bug.cgi?id=23967>
>
> By the way how do I actually run the tests when doing a CMake build? I
> don't see a nice target I can use to run them. I was thinking of maybe
> implementing this in a future PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825#issuecomment-115857246>.
>

abadams commented:
This seems to have broken on the build bot (which uses a downloaded llvm 3.5): 

https://drone.io/github.com/halide/Halide/1408

Am I doing anything wrong in that script?

abadams commented:
As far as I can tell, the llvm 3.5 release includes hardcoded paths in the cmake library to someone's dev box, and the llvm 3.6 release includes no cmake library at all :(

abadams commented:
Reverting for now to unbreak the build bot to unblock other stuff.

delcypher commented:
@abadams 

There are two issues I see with the llvm 3.5 release

* The hardcoded paths as you observed. I think this needs to fixed in upstream LLVM at some point
* The packager really screwed up here. Obviously they've either set CMAKE_INSTALL_PREFIX or  ``--prefix`` (Autoconf build system) to be ``/home/ben/development/llvm/3.5/final/Phase3/Release/llvmCore-3.5.0-final.install`` which is totally **wrong**. The install prefix should be ``/usr/local/`` and the ``DESTDIR`` makefile variable should be used when running ``make install`` to write the files elsewhere.

I believe the [LLVM PPA's are correctly packaged](http://llvm.org/apt/). for 3.5 and 3.6. Given that the your ``drone.io`` machine is using Ubuntu I would advise using those instead of downloading the binaries from the LLVM website in the build bot.

delcypher commented:
@abadams I found the issue in LLVM's release script. I'll send them a patch so the packaging isn't so bad in the future.

abadams commented:
We'd like people to be able to download and use precompiled llvms. This is
particularly useful when people don't have root on the system they're
using, or have a different global llvm install they want to keep. So
breaking the cmake build for llvm's precompiled releases would be a
significant regression.

On Fri, Jun 26, 2015 at 2:11 PM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams> I found the issue in LLVM's release
> script. I'll send them a patch so the packaging isn't so bad in the future.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825#issuecomment-115884301>.
>

delcypher commented:
@abadams Okay. I think I may have to give up on the idea of using LLVM's exported CMake targets. Clearly this feature isn't really ready for the prime time. I very much like the feature but it isn't worth breaking the existing behaviour.

When I find time I will redo this pull request using the old way of finding the LLVM header files, includes and binaries.

delcypher commented:
@abadams I've been looking into getting the CMake stuff in LLVM so that this pull request can be used. I'd rather not throw away what I've done just because the LLVM binaries have not been packaged properly (the Windows binaries should be okay because the CMake files generated by CMake are relocatable).

I have several patches for LLVM that will go into trunk soon that will make the CMake files generated by their Autoconf/Makefile build system (used to generate the released binaries) relocatable (i.e. no hardcoded paths). LLVM 3.6.2 is currently is release candidate phase and I've asked about getting these patches into the branch (I've not heard anything back yet).

If this happens then when 3.7 is released we will have three LLVM versions (I believe that is the current requirement) that we can target

* 3.8svn
* 3.7 release
* 3.6.2 release

These will all contain the necessary patches that will mean that the CMake files that come from the corresponding binary releases should be usable by my changes to Halide. This means that the changes in this pull request could be taken without modification.

What do you think?

zvookin commented:
I would be ok with more back rev versions of LLVM only being supported by
the Makefile or binary releases of Halide...

-Z-


On Mon, Jun 29, 2015 at 4:24 PM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams> I've been looking into getting the
> CMake stuff in LLVM so that this pull request can be used. I'd rather not
> throw away what I've done just because the LLVM binaries have not been
> packaged properly (the Windows binaries should be okay because the CMake
> files generated by CMake are relocatable).
>
> I have several patches for LLVM that will go into trunk soon that will
> make the CMake files generated by their Autoconf/Makefile build system
> (used to generate the released binaries) relocatable (i.e. no hardcoded
> paths). LLVM 3.6.2 is currently is release candidate phase and I've asked
> about getting these patches into the branch (I've not heard anything back
> yet).
>
> If this happens then when 3.7 is released we will have three LLVM versions
> (I believe that is the current requirement) that we can target
>
>    - 3.8svn
>    - 3.7 release
>    - 3.6.2 release
>
> These will all contain the necessary patches that will mean that the CMake
> files that come from the corresponding binary releases should be usable by
> my changes to Halide. This means that the changes in this pull request
> could be taken without modification.
>
> What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/825#issuecomment-116877946>.
>

delcypher commented:
@abestephensg @abadams

I've built on top this pull request ( https://github.com/delcypher/Halide/commits/dev ) and managed to get quite a few of the apps to build (Linux only, I haven't tested other platforms) :) .In doing this  I've found two bugs though relating to bad header file generation which I'm surprised no one has hit yet. More fun for tomorrow I guess.

delcypher commented:
@abestephensg @abadams

Well it looks like the upcoming LLVM 3.6.2 release will not have the necessary fix (r241080) to make the generated LLVM CMake files relocatable because we missed the merge window. It's not a decision I agree with (it's really a packaging change and does not effect the binaries) but there's nothing I can do.

On the plus side I discovered why the Ubuntu binary builds were broken. The ``test-release.sh`` script contained a pretty bad bug where the script would continue executing even if the build/tests failed. I've fixed that in r241599. The Ubuntu binaries for the upcoming LLVM 3.6.2 should now be less broken than last time (but still broken).

As a result using using the CMake files probably needs to postponed for a while. When I find time I will take the other useful changes (some will need to be reworked) that were in this PR and make a new PR.
